8190907: Windows 10 default Korean Font Malgun Gothic available not used by GUI elements even though available without Korean Language Pack

prsadhuk opened the pull request:
Malgun_Gothic(https://en.wikipedia.org/wiki/Malgun_Gothic) had replaced Gulim korean font so it needs to be updated to show korean fonts

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8190907](https://bugs.openjdk.org/browse/JDK-8190907): Windows 10 default Korean Font Malgun Gothic available not used by GUI elements even though available without Korean Language Pack


### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to [1b7a5993](https://git.openjdk.org/jdk/pull/7643/files/1b7a5993c409793e244fd7cd28e97bb9a816834f)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/7643/head:pull/7643` \
`$ git checkout pull/7643`

Update a local copy of the PR: \
`$ git checkout pull/7643` \
`$ git pull https://git.openjdk.org/jdk pull/7643/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7643`

View PR using the GUI difftool: \
`$ git pr show -t 7643`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/7643.diff">https://git.openjdk.org/jdk/pull/7643.diff</a>

</details>

magicus commented:
/label -build

prsadhuk commented:
Malgun Gothic is supposed to be the default korean font replacing Gulim as per the wiki mentioned in the description, so I have only replaced Gulim,not all.. Others like Batang are not replaced.
Also, as I pointed out by screenshot in JBS in my Windows10 I dont have malgunbd.ttf so I was doubting it may not be available in all distribution by default, so I have used only malgun.ttf

OTOH, if Gulim is available as supplementary fonts, we probably can close this issue as "Wont fix" mentioning the same.

naotoj commented:
I thought there is some mechanism to chain physical fonts, such as if `glim.ttc` is not available for `korean` then use `malgun.ttf` instead, but I may be wrong here. @prrace, can you comment on it?

If there is no such fallback mechanism, `Batang` should also be replaced with `Malgun Gothic`, otherwise it would show blobs.

prrace commented:
There's no chaining mechanism in the way you describe it .. it would have to be in our fallback list.

Batang and Gulim are the fonts that have been available since .. well at least XP.

https://docs.microsoft.com/en-us/typography/font-list/batang

https://docs.microsoft.com/en-us/typography/font-list/gulim

Malun Gothic only came in with Windows 7 :-
https://docs.microsoft.com/en-us/typography/font-list/malgun-gothic

I think it reasonable to put malun.ttf in the "allfonts" as a fallback  if it really is pre-installed on ALL systems.

But actual Korean users really should be installing the Korean font pack regardless and then
we should pick the best available font. Maybe its the new ones now.

Replacing them would not have been possible in the past but now Vista isn't supported

Note that Gulim is a Sans Serif font, Batang a Serif font.
Malun Gothic is described as suitable for on screen display - it sounds like a replacement for Gulim
but not for Batang so it is correct you aren't replacing that.
But is there a new Serif font available too ?
Please investigate / check.

And it seems (ah I see Naoto spotted this too) there is a Bold version of
Malgun Gothic in its own ttf file : Malgunbd.ttf So we should start using that.
To your point about it not being in the pre-installed fonts it doesn't matter for the reason I gave.

One other thing you are going to have to check is the rendering of other scripts when running in a Korean locale.

Batang lists support for Arabic .. Malun does not .. but then neither does Gulim .. so it may be no worse.

But sequence.fallback doesn't have any obvious Arabic support.
We used to be able to count on Lucida Sans for that .. but may be there's another gap there.
Mainly make sure you don't increase that gap.

prsadhuk commented:
> But is there a new Serif font available too ?

I could not find any replacement for Batang serif font so have kept it same.

> Bold version of Malgun Gothic in its own ttf file : Malgunbd.ttf So we should start using that

Added

> Arabic support.

Tested with java/awt/font/NumericShaper/EasternArabicTest and MTTest..Rendered same before and after this fix.

naotoj commented:
OK, so there is no way to chain the physical fonts.
One other thing I noticed was that `GulimChe` is a `monospaced` font, so I would also leave it for `monospaced` properties, as `Malgun Gothic` is a proportional font.

prrace commented:
> OK, so there is no way to chain the physical fonts.
> One other thing I noticed was that `GulimChe` is a `monospaced` font, so I would also leave it for `monospaced` properties, as `Malgun Gothic` is a proportional font.

Good point.

prrace commented on make/data/fontconfig/windows.fontconfig.properties:
```diff
@@ -144,56 +145,56 @@ dialog.plain.chinese-ms950=MingLiU
 dialog.bolditalic.hebrew=David Bold
 dialog.bolditalic.japanese=MS Gothic
-dialog.bolditalic.korean=Gulim
+dialog.bolditalic.korean=Malgun Gothic
```
You should use the Bold version of the font here too - ie for bold italic - not just for bold - since it is clearly closer ..

prrace commented on make/data/fontconfig/windows.fontconfig.properties:
```diff
@@ -144,56 +145,56 @@ dialog.plain.chinese-ms950=MingLiU
 dialoginput.bold.japanese=MS Gothic
-dialoginput.bold.korean=Gulim
+dialoginput.bold.korean=Malgun Gothic Bold
 
```
Since dialoginput is usually the same as monospaced, per Naoto's comment, this should be reverted.

naotoj replied:
Not only reverting, but I think `GulimChe` here is the right choice, as it is `monospaced`?

JB-Dmitry commented:
If I may comment, in the past we received a noticeable amount of complaints about Korean text not rendered in our IDEs on Windows. This was solved by installing Korean font pack, but relying on users always doing that by themselves doesn't seem to provide the best user experience. So, two years ago, in JetBrains Runtime we ended up replacing Gulim/GulimChe/Batang with Malgun Gothic. This works well for us so far. I understand that this approach can give worse results for users that do install Korean font pack. But, I believe, the 'chaining' proposed by Naoto can actually be implemented, if desired - we just need to introduce a new 'category' to the properties file, say `korean_supplementary`. By putting it before `korean` in sequence definitions, we'll make e.g. Batang used if it's installed, and fall back to Malgun Gothic otherwise.

naotoj commented:
Initially I thought it was surprising this problem has not been reported quite some time since `Malgun Gothic` was introduced. Then I guessed that Korean customers possibly would all install Korean language pack because without it, they would not be able to input Korean chars (no Korean input method included by default). I think in IntelliJ's case, possibly the typical users are all developers installing English Windows, thus there have been bug reports. So, I would rather think replacing those fonts with `Malgun Gothic` could be problematic to the former Korean customers.
Having said that, I would want the `chaining` functionality which could effectively satisfy both users. Other than the option you provided, there could be an extension of the fontconfig, such as,
```
serif.plain.korean=Batang, Malgun Gothic
```
so that if `Batang` file is not ther, we can fall back to `Malgun Gothic`.

prrace commented:
But serif.plain.korean isn't used unless it is a Korean locale .. so what's the point ?

And this one line already added 
allfonts.korean=Malgun Gothic
will provide it as a fall back so I don't see the need.
If I am an English language user running in a US / UK locale I'll get my Korean glyphs from there with the only difference being it isn't in the same place in the precedence order.
This also ensures that any "new" code points added in that font not present in the older ones will always be available.

naotoj commented:
> allfonts.korean=Malgun Gothic will provide it as a fall back so I don't see the need.

OK, that's what I needed. Thanks.

prrace commented:
BTW does anyone know if this is a problem for Chinese or Japanese too ?
I think Indic might have needs a font pack too .. maybe not in this bug but should we be looking wider than Korean ?

naotoj commented:
Not that I know of. Simple testing with Japanese text worked OK, but extensive testing for CJK would be needed.

JB-Dmitry commented:
> But serif.plain.korean isn't used unless it is a Korean locale

That's not the case, and is easily verifiable for English locale - whatever font is specified there will become a component of Serif composite. 'korean' here is a script id, which is referenced from `sequence.fallback`, used for all locales.

> And this one line already added
allfonts.korean=Malgun Gothic
will provide it as a fall back so I don't see the need.

That's also not the case. With the patch applied, Serif font still doesn't have Malgun Gothic as a component.
That line in the current patch has no effect at all - `allfonts.<script>` lines only have an effect for scripts, which don't have assigned fonts in `<composite>.<style>.<script>` properties. 

> BTW does anyone know if this is a problem for Chinese or Japanese too ?
I think Indic might have needs a font pack too .. maybe not in this bug but should we be looking wider than Korean ?

Didn't hear about related problems for Chinese or Japanese, but the same problem exists for Thai script - DokChampa font isn't available by default on Windows 10, and we use Tahoma as a replacement currently.

BTW, maybe it's worth to create a test case for this problem? As Korean support is explicitly coded in `windows.fontconfig.properties`, it might be good to check that all 20 logical fonts can display Korean on Windows. If a newer Windows version will change the bundled font again, the problem will become apparent at once.

prrace commented:
I'm reviewing https://docs.oracle.com/javase/8/docs/technotes/guides/intl/fontconfig.html#windows where it says
The keys have the forms:

        allfonts.CharacterSubsetName
        LogicalFontName.StyleName.CharacterSubsetName
The first form is used if the same font is used for a character subset independent of logical font and style (in this case, the font rendering engines apply algorithmic styles to the font). The second form is used if different physical fonts are used for a character subset for different logical fonts and styles. In this case, properties must be specified for each combination of logical font and style, so 20 properties for one character subset. If a property of the first form is present for a character subset, then properties of the second form for the same character subset are ignored.

So if you remove the allfonts entry, does it pick up serif.plain.korean (etc) ?

And the allfonts entry does (I am fairly sure) then apply to the fallback sequence if there are no "second form" entries, but regardless it seems odd that fallback entries are ignored because the second form is present, but ignored and in turn causing the fallback to be ignored ? 

This does not seem to be working intuitively.
Perhaps it hasn't been much of an issue if serif.plain.korean (etc) generally refers to an existent font that is usually installed and no one needs to worry about the fallback entries. But I think nothing should "hide" fallback entries.

So this may need code changes as well. And that doc may need to be updated ..

prrace commented:
Hmm .. I have not tried this but a "hack" might be to replace
korean.allfonts
with
korean-fallback.allfonts
and similarly update sequence.fallback to refer to korean-fallback

It is all just name matching so korean could be "foo" and if the duplicate use is the issue that could fix it.

prrace commented:
I just did a quick hack test and it looks like it might do it.

prrace commented:
And @JB-Dmitry what did you actually test ? The current PR state ? 
I assume that @prsadhuk tested that before updating the PR.
But the current PR state looks like what you say doesn't work so I am getting confused.

JB-Dmitry commented:
> So if you remove the allfonts entry, does it pick up serif.plain.korean (etc) ?

JDK picks up `<LogicalFontName>.<StyleName>.korean` entries regardless of whether `allfonts.korean` entry is present - the former take precedence over the latter. Yes, it goes against the phrase "If a property of the first form is present for a character subset, then properties of the second form for the same character subset are ignored." in documentation, but that's how code in `sun.awt.FontConfiguration.PropertiesHandler.initBinaryTable()` currently works. Indeed it would be good to fix either the code or the documentation. 

> And the allfonts entry does (I am fairly sure) then apply to the fallback sequence if there are no "second form" entries, but regardless it seems odd that fallback entries are ignored because the second form is present, but ignored and in turn causing the fallback to be ignored ?

Not sure what you mean by fallback entries here. If it's about `sequence.fallback` entries, they aren't ignored, they just use fonts specified by the "second form", not the "first form".

> And @JB-Dmitry what did you actually test ? The current PR state ?

Yes, I tested the latest PR state. Just like the initial PR version, and current code in master, in that state Serif font will display boxes for Korean text, if Korean font pack isn't installed, at least on my Windows 10 machine.

prsadhuk commented:
> I assume that @prsadhuk tested that before updating the PR.

I tested with the testcase attached with the bug...not sure of the Serif usage..

![image](https://user-images.githubusercontent.com/43534309/157863549-06e83424-99b5-4cff-9714-ae8ce7f573a9.png)

prsadhuk commented:
I tested with this test and I got the following output which seems to suggest "serif" korean is working...
```
BufferedImage bi = new BufferedImage(400, 200, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = bi.createGraphics();
        g.setColor(Color.WHITE);
        g.fillRect(0,0,400,200);
        g.setColor(Color.BLACK);
        g.setFont(new Font("sansserif", Font.PLAIN, 20));
        g.drawString("테스트", 100, 150);
        g.setFont(new Font("serif", Font.BOLD, 20));
        g.drawString("테스트", 100, 100);
        ImageIO.write(bi, "png", new File("image.png"));
```
![image](https://user-images.githubusercontent.com/43534309/157869281-caece0c1-6a79-4a0a-93dc-bec9565709be.png)

prsadhuk commented:
> @prsadhuk Do you have Korean font pack installed? Batang font, in particular.

I tested with both Korean pack installed and after uninstalling it...But I have batang.ttc even when I removed korean font pack..

prrace commented:
>> And the allfonts entry does (I am fairly sure) then apply to the fallback sequence
>> if there are no "second form" entries, but regardless it seems odd that fallback
>> entries are ignored because the second form is present, but ignored and in turn
>> causing the fallback to be ignored ?

> Not sure what you mean by fallback entries here. If it's about sequence.fallback entries,
> they aren't ignored, they just use fonts specified by the "second form", not the "first form".

Yes, the sequence.fallback but I am relaying my observation that adding
korean.allfonts=..
did not work even though sequence.fallback lists korean whereas adding
korean-foo.allfonts=..
and adding korean-foo into sequence.fallback DID work.

So the "first form" being present in the non-working case is the only difference and infer that it hides the usage of allfonts by the fallback sequence.

Someone would need to follow through the code to see why.

And I did not have the korean font pack installed.

Probably this hiding should be regarded as a bug unless we look and see a good reason for it, but it isn't obvious or documented.

JB-Dmitry commented:
> Yes, the sequence.fallback but I am relaying my observation that adding
korean.allfonts=..
did not work even though sequence.fallback lists korean whereas adding
korean-foo.allfonts=..
and adding korean-foo into sequence.fallback DID work.

It's just there are (only) two ways to define Korean (or any other script) font to be used for fallback - either use only one of them for all 20 logical Java fonts (like e.g. it's currently done for Thai language), or configure a different Korean font for each of those 20 fonts separately (like it's currently done in OpenJDK for Korean language). It's not possible to 'mix' those two approaches - by using 'allfonts' entry as a kind of 'fallback' for the case when a specific font doesn't exist. When both ways (forms) are provided in the configuration file, current code picks the 'second' form, ignoring `allfonts` entry. That's why adding `allfonts.korean` in the patch didn't have any effect. But there are no existing entries like `serif.plain.korean-foo`, so adding `allfonts.korean-foo` does have an effect.

JB-Dmitry commented:
It's a bit hard to talk about 'fallback' fonts here without confusion, as this word might mean at least two things - we have Korean fallback fonts, in a sense that `sequence.fallback` contains `korean` entry, so all logical fonts are supposed to support Korean. And we also want e.g. Batang font to 'fallback' to Malgun Gothic, if Batang font is not installed.

prsadhuk commented:
> It's a bit hard to talk about 'fallback' fonts here without confusion, as this word might mean at least two things - we have Korean fallback fonts, in a sense that `sequence.fallback` contains `korean` entry, so all logical fonts are supposed to support Korean. And we also want e.g. Batang font to 'fallback' to Malgun Gothic, if Batang font is not installed.

It seems then if we have
`allfonts.korean-fallback=Malgun Gothic`
and add this `korean-fallback` to `sequence.fallback` instead of `korean` then I guess we can have fix for both.
Will this be accepted change?

prrace commented:
> It's a bit hard to talk about 'fallback' fonts here without confusion, as this word might mean at least two things - we have Korean fallback fonts, in a sense that `sequence.fallback` contains `korean` entry, so all logical fonts are supposed to support Korean. And we also want e.g. Batang font to 'fallback' to Malgun Gothic, if Batang font is not installed.

I don't see why it is hard. In fact there's actually only one thing.
The fallback sequence is appended to all logical fonts of all styles.
It means that if a font that is specified in a particular logical font isn't present, something from this sequence will provide the fallback.
There's no font or script specific fallback logic along the lines of if (!batang) then insert malgun in its same place.
It is really code point fallback : if asked for codepoint \u1101, then search the component fonts in order until you find it or run out of fonts. 

There's some unintended behaviour (bug) that "foo" in the logical fonts is hiding "foo" in the fallback sequence but that doesn't change the above logic.

prrace commented:
> It seems then if we have
allfonts.korean-fallback=Malgun Gothic
and add this korean-fallback to sequence.fallback instead of korean then I guess we can have fix for both.
Will this be accepted change?

It is fine by me. I just want to be hear that you have tested all the scenarios and it works as we want it to.

prsadhuk commented:
> > It seems then if we have
> > allfonts.korean-fallback=Malgun Gothic
> > and add this korean-fallback to sequence.fallback instead of korean then I guess we can have fix for both.
> > Will this be accepted change?
> 
> It is fine by me. I just want to be hear that you have tested all the scenarios and it works as we want it to.

I have tested with sanserif, serif, monospaced, dialog, dialoginput with PLAIN, BOLD, ITALIC, BOLD-ITALIC style in that order which is rendered correctly
![image](https://user-images.githubusercontent.com/43534309/158526001-8dfb8fb7-ffdd-4ded-8f56-714b0680d998.png)

prrace reviewed (approved):
The image looks like it is using Malgun for all fonts, so I'd like to see one with the font pack installed too.

prsadhuk commented:
With korean font pack installed, language set to Korean and calling Locale.setDefault(Locale.KOREAN) in testcase, we have
![image](https://user-images.githubusercontent.com/43534309/158734779-78e72fd2-e8e1-4991-a332-cd1624692282.png)

prrace commented:
> With korean font pack installed, language set to Korean and calling Locale.setDefault(Locale.KOREAN) in testcase, we have
> ![image](https://user-images.githubusercontent.com/43534309/158734779-78e72fd2-e8e1-4991-a332-cd1624692282.png)

And if you revert your entire change and KEEP the Korean font pack ? What does the entire image look like ?
Because both your images look the same .. which isn't what I'd expect.
Basically SHOW there is a difference in one of the fonts that is replaced so I can be sure it is working.

prsadhuk commented:
Without the fix,
![image](https://user-images.githubusercontent.com/43534309/158740896-b0ee2daf-b8c3-4dc9-9f3b-7a1831562a12.png)

prrace commented:
So clearly there is still something wrong with the fix. Some of these should not have changed. You'd see it even more clearly if you labelled each of the pieces of text with the font. I don't know what. I'm out of suggestions. It requires debugging which is beyond the scope of reviewing a proposed fix. So you really need to test it is doing what is expected and "something displays" is not enough. It has to be the expected and right something.
So please do that and come back with something that resolves all these problems.

prsadhuk commented:
> Without the fix, ![image](https://user-images.githubusercontent.com/43534309/158740896-b0ee2daf-b8c3-4dc9-9f3b-7a1831562a12.png)

If I compile the testcase with in-built javac and then run with build jdk, then I get the above image
but if I compile testcase with build javac (as in ./jdk/bin/javac) and then run with build jdk then I get (This is without the change in fontconfig)
![image](https://user-images.githubusercontent.com/43534309/158744766-1cef1101-3738-4315-a421-6069a3840dff.png)

prrace commented:
It seems that the test program is not using \uXXXX unicode escapes but has the characters in some encoding that needs to be expected by javac.
So the first image is from a javac that was using default (eg US) locale and the latter was using a locale that can correctly interpret those.

I don't know why the two are different for you - probably it is a shell environment difference - but I recommend you always use \uXXXX in tests to avoid exactly this issue.

Now looking at the 2nd image it shows that there are differences (although again it would be nice if each line were labelled - eg "Dialog Bold Italic : " ...) and so something is not working as intended.

I don't know what the problem is, so it will require some extensive debugging.

prsadhuk commented:
I was not sure how to use "테스트" korean (which was there in JBS testcase) in \uXXX unicode so I used
g.drawString("테스트", 100, 80);

prrace reviewed (changes requested):
Not ready yet

prsadhuk commented:
As I mentioned initially,
> tested with sanserif, serif, monospaced, dialog, dialoginput with PLAIN, BOLD, ITALIC, BOLD-ITALIC style **in that order**

so 1st 4 are sansserif, next 4 are serif etc.

As I see, we have this font support in the fontconfig properties 
```
Serif = Batang				 batang.ttc
sansserif = Malgun Gothic	malgun.ttf
monospaced = Gulimche		gulim.ttc
dialog = Malgun Gothic		malgun.ttf
dialoginput = Gulim			gulim.ttf
```
Now in my WIndows 10, I have these font files irrespective of Korean font pack installed or not, 
so I guess that is why there is **no difference** in the font image with the fix with/without korean font pack



Now, without the fix, [comparing with/without the fix with Korean font pack installed https://github.com/openjdk/jdk/pull/7643#issuecomment-1070307750 and https://github.com/openjdk/jdk/pull/7643#issuecomment-1070348395 ] there is minor difference (AFICS) in bold, italic corresponds to sansserif (2-3 position), monospaced (10-11), dialog (14-15), dialoginput (18-19) which used Gulim before fix but used Malgun Gothic Bold with fix **which i guess is expected**..

Also, there is clear difference in 6-8 th position for Bold, Italic, Bold-Italic which corresponds to **serif font** which uses same Batang font before and after the fix which I guess is because of the "first form", "second form" fallback issue discusses in this thread . I think with the fix `allfont.korean-fallback` is in effect so it uses malgun (which is why 1-4 position looks same as 5.-8) and without fix, it uses Batang

prrace commented:
I'm confused since your images with and without the font pack look the same yet they
should be using different fonts - ie Gulimche should be used for monospaced when you have the font pack - although maybe  it is always installed, but also  the rendering of each group looks the same so serif == sanserif == monospaced == dialoginput == dialog - why ?

Which should not be the case at least with the font pack

prsadhuk commented:
> I'm confused since your images with and without the font pack look the same yet they should be using different fonts - ie Gulimche should be used for monospaced when you have the font pack - although maybe it is always installed, but also the rendering of each group looks the same so serif == sanserif == monospaced == dialoginput == dialog - why ?
> 
> Which should not be the case at least with the font pack

Are you asking this "with the fix"? As I mentioned, 

> Now in my WIndows 10, I have these font files irrespective of Korean font pack installed or not,
> so I guess that is why there is no difference in the font image **with the fix** with/without korean font pack

prrace commented:
I still don't understand why they are the same for you .. I added korean-fallback to my local fontconfig and ran in the Korean locale (changed my windows region to Korea and rebooted) and they are different for me.

What I do seem to be seeing is that because we now have UTF-8 as the default encoding and that we have no 
UTF-8.ko that it goes the fallback route .. and so although you get korean by fallback it prefers Chinese and Japanese fonts which isn't right for a Korean locale

I believe we need to add a line like the ja one 
sequence.allfonts.UTF-8.ja=alphabetic,japanese,dingbats,symbol
+sequence.allfonts.UTF-8.ko=alphabetic,korean,dingbats,symbol

[See https://git.openjdk.java.net/jdk/pull/8662 for more info and we likely need to do more of these lines and perhaps you need that PR pushed to include in your testing .. alas I don't even have a single approver yet, so nowhere near the two I need !]

.. even without that I was OK perhaps so long as the fallback seqeuence had "korean" - I'd have to re-test that

So I don't know if that will cure  them all being the same for you but it might.

prsadhuk commented:
> I still don't understand why they are the same for you .. I added korean-fallback to my local fontconfig and ran in the Korean locale (changed my windows region to Korea and rebooted) and they are different for me.
> 
> What I do seem to be seeing is that because we now have UTF-8 as the default encoding and that we have no UTF-8.ko that it goes the fallback route .. and so although you get korean by fallback it prefers Chinese and Japanese fonts which isn't right for a Korean locale
> 
> I believe we need to add a line like the ja one sequence.allfonts.UTF-8.ja=alphabetic,japanese,dingbats,symbol +sequence.allfonts.UTF-8.ko=alphabetic,korean,dingbats,symbol
> 
> [See https://git.openjdk.java.net/jdk/pull/8662 for more info and we likely need to do more of these lines and perhaps you need that PR pushed to include in your testing .. alas I don't even have a single approver yet, so nowhere near the two I need !]
> 
> .. even without that I was OK perhaps so long as the fallback seqeuence had "korean" - I'd have to re-test that
> 
> So I don't know if that will cure them all being the same for you but it might.

I tried the same way of changing region to Korean and rebooting and tried this testcase (changed keyboard IM to Korean too)
I have also added sequence.allfonts.UTF-8.ko=alphabetic,korean,dingbats,symbol

```
public class HelloWorld {
    private static void createAndShowGUI() {
         JFrame frame = new JFrame("테스트 Swing");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);

        JLabel label = new JLabel("테스트");
        frame.getContentPane().add(label);

        frame.setSize(400,200);
        frame.setVisible(true);
    }

    public static void main(String[] args) throws Exception {
        Locale.setDefault(Locale.KOREAN);
        javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
                createAndShowGUI();
            }
        });
        BufferedImage bi = new BufferedImage(600, 900, BufferedImage.TYPE_INT_RGB);
        Graphics2D g = bi.createGraphics();
        g.setColor(Color.WHITE);
        g.fillRect(0,0,600,900);
        g.setColor(Color.BLACK);

        g.setFont(new Font("sansserif", Font.PLAIN, 20));
        g.drawString("Sansserif Plain:  테스트", 100, 40);

        g.setFont(new Font("sansserif", Font.BOLD, 20));
        g.drawString("Sansserif Bold:  테스트", 100, 80);

        g.setFont(new Font("sansserif", Font.ITALIC, 20));
        g.drawString("Sansserif Italic:  테스트", 100, 120);

        g.setFont(new Font("sansserif", Font.BOLD|Font.ITALIC, 20));
        g.drawString("Sansserif BoldItalic:  테스트", 100, 160);

        g.setFont(new Font("serif", Font.PLAIN, 20));
        g.drawString("Serif Plain:  테스트", 100, 200);

        g.setFont(new Font("serif", Font.BOLD, 20));
        g.drawString("Serif Bold:  테스트", 100, 240);

        g.setFont(new Font("serif", Font.ITALIC, 20));
        g.drawString("Serif Italic:  테스트", 100, 280);

        g.setFont(new Font("serif", Font.BOLD|Font.ITALIC, 20));
        g.drawString("Serif BoldItalic:  테스트", 100, 320);

        g.setFont(new Font("monospaced", Font.PLAIN, 20));
        g.drawString("Monospaced Plain:  테스트", 100, 360);

        g.setFont(new Font("monospaced", Font.BOLD, 20));
        g.drawString("Monospaced Bold:  테스트", 100, 400);

        g.setFont(new Font("monospaced", Font.ITALIC, 20));
        g.drawString("Monospaced Italic:  테스트", 100, 440);

        g.setFont(new Font("monospaced", Font.BOLD|Font.ITALIC, 20));
        g.drawString("Monospaced BoldItalic:  테스트", 100, 480);

        g.setFont(new Font("dialog", Font.PLAIN, 20));
        g.drawString("Dialog Plain:  테스트", 100, 520);

        g.setFont(new Font("dialog", Font.BOLD, 20));
        g.drawString("Dialog Bold:  테스트", 100, 560);

        g.setFont(new Font("dialog", Font.ITALIC, 20));
        g.drawString("Dialog Italic:  테스트", 100, 600);

        g.setFont(new Font("dialog", Font.BOLD|Font.ITALIC, 20));
        g.drawString("Dialog Bold Italic:  테스트", 100, 640);

        g.setFont(new Font("dialoginput", Font.PLAIN, 20));
        g.drawString("DialogInput Plain:  테스트", 100, 680);

        g.setFont(new Font("dialoginput", Font.BOLD, 20));
        g.drawString("DialogInput Bold:  테스트", 100, 720);

        g.setFont(new Font("dialoginput", Font.ITALIC, 20));
        g.drawString("Dialog Input Italic:  테스트", 100, 760);

        g.setFont(new Font("dialoginput", Font.BOLD|Font.ITALIC, 20));
        g.drawString("Dialog Input BoldItalic:  테스트", 100, 800);

        ImageIO.write(bi, "png", new File("image.png"));
    }
}
```

and I am getting this rendered fonts
![image](https://user-images.githubusercontent.com/43534309/170485724-51909f48-db09-4a4f-bec4-61c78e5baff7.png)

prsadhuk commented:
I used unicode characters (\ud14c\uc2a4\ud2b8) in testcase and run with user.language=ko and user.country=KR and now I can see the difference between Dialog and Serif
![image](https://user-images.githubusercontent.com/43534309/170640958-70d6b34d-63e6-475e-8321-b1723fc6438e.png)

prrace commented:
Good to know. I still believe we should add
sequence.allfonts.UTF-8.ko=alphabetic,korean,dingbats,symbol 
since UTF-8 is now the default encoding.

prsadhuk commented:
> Good to know. I still believe we should add sequence.allfonts.UTF-8.ko=alphabetic,korean,dingbats,symbol since UTF-8 is now the default encoding.

Added UTF-8 korean fallback. I get the same above rendering with the same commandline as above..

prrace commented:
Something weird is going on .. 
Malgun is now not appearing in the font menu of Font2DTest when I use  ko + UTF-8 !?
It appears if I use -Dfile.encoding=ko instead.

Something we are doing is hiding it.

Also can you attach the source of your test to JBS !

prsadhuk commented:
> Also can you attach the source of your test to JBS !

I have attached.

prsadhuk commented:
> Malgun is now not appearing in the font menu of Font2DTest when I use ko + UTF-8 !?
> It appears if I use -Dfile.encoding=ko instead

Are you implying it used to appear earlier? with what configuration or patch?
It is same for me but I haven't run Font2DTest earlier with this patch so could not say if it used to work without file.encoding=ko. 
It seems to not appear even with -Duser.language=ko -Duser.country=KR -Dfile.encoding=ko. I also tried assing "ko" to 8285308  fix but to no avail.

prsadhuk commented:
I see even `Gulimche `and `Batang `are not visible without -Dfile.encoding=ko so maybe it's what is expected.

prsadhuk commented:
> So Malgun IS listed. But when I am running in a fully Korean configuration, Font2DTest is displaying the Korean name for the font - not the English name. So oops, all was just fine. This fix seems ready to push

Hopefully we dont have to sit with interpreter in future for CJK fixes :-)

prsadhuk commented:
/integrate
